### PR TITLE
Write DLQ entries to temp file first

### LINKED ISSUE #8022: DLQ: new files on every restart
When the Dead Letter Queue (DLQ) is enabled, every restart or pipeline restart of Logstash creates a new tiny file that may never be used.  

For example, I restarted the pipeline 9 times, and sent 0 events:

 ls data/dead_letter_queue/main | wc -l
`9`

The content of these files:
cat data/dead_letter_queue/main/1.log
`1`

I am not sure if there is a reaper that would eventually clean this up, but it seems unnecessary to pre-create a file for every pipeline start.

This can result in confusion, and the appearance of an error if watching the metrics. Since each file has a size, it artificially inflates the dead letter queue size. When I first saw this, i thought data was actually going into the DLQ, but that was not the case.

```
      "dead_letter_queue" : {
        "queue_size_in_bytes" : 9
      }
```

cc @robbavey

### PR DESCRIPTION
**Pull request by robbavey**
This commit changes the DLQ writer to write to a temporary file
 which will be renamed on "completion", to avoid the possibility
 of the DLQ reader reading an incomplete DLQ segment. The temp file
 will be renamed and made available, either when the capacity of this
 segment is reached, or if a configurable 'flush interval' has elapsed
 since the last event reached the dead letter queue.

This commit fixes #8022, #10275, #10967

### CODE REVIEW

**Comment by robbavey** [comment]
jenkins test this please

**Comment by amitavmohanty01** [comment]
In this approach, the system needs to manage scenarios like: what happens when the process crashes before the temporary file is renamed.

An alternative approach: the system already maintains a marker of sorts with the `sincedb` feature. There could be another marker indicating `safe-read-upto` value. The writer keeps writing freely and updates this marker indicating the segment to which the reader can safely read.

I am just putting it out for discussion.

File: qa/integration/specs/dlq_spec.rb [hunk 1]
```diff
@@ -68,18 +75,20 @@
...
     context 'with multiple pipelines' do
+      let(:message) { generate_message(100)}
       let(:pipelines) {[
           {
               "pipeline.id" => "test",
               "pipeline.workers" => 1,
               "dead_letter_queue.enable" => true,
               "pipeline.batch.size" => 1,
-              "config.string" => "input { generator { message => '{\"test\":\"one\"}' codec => \"json\" count => 1000 } } filter { mutate { add_field => { \"geoip\" => \"somewhere\" } } } output { elasticsearch {} }"
+              "config.string" => "input { generator { message => '#{message}' codec => \"json\" count => 1000 } } filter { mutate { add_field => { \"geoip\" => \"somewhere\" } } } output { elasticsearch {} }"
```
Comment (amitavmohanty01) [inline qa/integration/specs/dlq_spec.rb]: It does not matter in this case but string extrapolation works with double quoted strings in ruby.

**Comment by renshuki** [comment]
@robbavey is there any blocker that prevents this PR to get merged? Thanks.

**Pull request closed without merging**